File: internal/agentrun/intelligence_test.go
package agentrun

import (
	"testing"
	"time"

	"github.com/ro-ag/ptrack/internal/association"
)

func intelligenceRun() Run {
	return Run{
		ID: "run-1", Provider: "codex", Kind: RegistrationExternal,
		State: StateRunning, ProcessState: ProcessUnknown, LeaseState: LeaseActive,
		ProjectRoot: "/project", CWD: "/project",
	}
}

func intelligenceEvent(sequence uint64, kind EventKind, phase EventPhase) Event {
	return Event{
		ModelVersion: EventModelVersion, ID: "event-" + string(rune('0'+sequence)),
		RunID: "run-1", Provider: "codex", HostSequence: sequence,
		Kind: kind, Phase: phase, ObservedAt: time.Unix(int64(sequence), 0),
		Correlation: EventCorrelation{ProjectRoot: "/project"},
	}
}

func TestDeriveRunIntelligenceDecisionTable(t *testing.T) {
	tests := []struct {
		name       string
		mutateRun  func(*Run)
		events     []Event
		state      IntelligenceState
		confidence IntelligenceConfidence
	}{
		{name: "live evidence floor is working not waiting", state: IntelligenceWorking, confidence: ConfidenceLow},
		{name: "explicit waiting", events: []Event{intelligenceEvent(1, EventLifecycle, EventWaiting)}, state: IntelligenceWaiting, confidence: ConfidenceMedium},
		{name: "explicit blocked", events: []Event{intelligenceEvent(1, EventTool, EventBlocked)}, state: IntelligenceBlocked, confidence: ConfidenceMedium},
		{name: "explicit completion", events: []Event{intelligenceEvent(1, EventLifecycle, EventCompleted)}, state: IntelligenceCompleted, confidence: ConfidenceHigh},
		{name: "explicit lifecycle failure", events: []Event{intelligenceEvent(1, EventLifecycle, EventFailed)}, state: IntelligenceFailed, confidence: ConfidenceHigh},
		{
			name: "nonzero exit", mutateRun: func(run *Run) {
				run.State = StateExited
				run.LeaseState = LeaseExpired
				run.Exit = &Exit{Code: 2, Result: "failed"}
			},
			events: []Event{intelligenceEvent(1, EventLifecycle, EventCompleted)},
			state:  IntelligenceFailed, confidence: ConfidenceHigh,
		},
		{
			name: "successful exit alone is unknown", mutateRun: func(run *Run) {
				run.State = StateExited
				run.LeaseState = LeaseExpired
				run.Exit = &Exit{Code: 0, Result: "completed"}
			},
			state: IntelligenceUnknown,
		},
		{
			name:   "test failure is not run failure",
			events: []Event{intelligenceEvent(1, EventTest, EventFailed)},
			state:  IntelligenceWorking, confidence: ConfidenceMedium,
		},
		{
			name: "stale silence is unknown", mutateRun: func(run *Run) {
				run.State = StateStale
				run.LeaseState = LeaseExpired
			},
			state: IntelligenceUnknown,
		},
		{
			name: "stale waiting evidence is not a current wait", mutateRun: func(run *Run) {
				run.State = StateStale
				run.LeaseState = LeaseExpired
			},
			events: []Event{intelligenceEvent(1, EventLifecycle, EventWaiting)},
			state:  IntelligenceUnknown,
		},
		{
			name: "stale blocked evidence is not a current block", mutateRun: func(run *Run) {
				run.State = StateStale
				run.LeaseState = LeaseExpired
			},
			events: []Event{intelligenceEvent(1, EventLifecycle, EventBlocked)},
			state:  IntelligenceUnknown,
		},
	}
	for _, test := range tests {
		t.Run(test.name, func(t *testing.T) {
			run := intelligenceRun()
			if test.mutateRun != nil {
				test.mutateRun(&run)
			}
			got := DeriveRunIntelligence(run, test.events)
			if got.State != test.state || got.Confidence != test.confidence {
				t.Fatalf("intelligence = %#v, want state=%q confidence=%q", got, test.state, test.confidence)
			}
		})
	}
}

func TestRegistryIntelligenceAppliesRetentionBeforeDerivation(t *testing.T) {
	now := time.Date(2026, time.August, 10, 12, 0, 0, 0, time.UTC)
	clock := &fakeClock{now: now}
	policy := DefaultEventPrivacyPolicy()
	policy.RetainFor = time.Second
	registry := newEventRegistryForTest(t, Config{
		ProjectRoot: "/project", LeaseDuration: time.Hour,
		Now: clock.Now, NewTicker: clock.NewTicker, EventPolicy: &policy,
	})
	lease, err := registry.RegisterExternal(Registration{
		Profile: "wrapper", Provider: "codex", CWD: "/project",
	})
	if err != nil {
		t.Fatal(err)
	}
	if _, err := registry.RecordEvent(lease.Run.ID, lease.LeaseToken, EventObservation{
		ModelVersion: EventModelVersion, SourceID: "wait-1", SourceSequence: 1,
		Kind: EventLifecycle, Phase: EventWaiting,
	}); err != nil {
		t.Fatal(err)
	}
	clock.Advance(2 * time.Second)
	intelligence, err := registry.Intelligence(lease.Run.ID)
	if err != nil {
		t.Fatal(err)
	}
	if intelligence.State != IntelligenceWorking ||
		intelligence.Confidence != ConfidenceLow || intelligence.EventCount != 0 {
		t.Fatalf("expired evidence intelligence = %#v", intelligence)
	}
}

func TestDeriveRunIntelligenceRequiresCurrentExplicitDriftEvidence(t *testing.T) {
	run := intelligenceRun()
	run.Association = &association.AssociationV1{
		Version: association.VersionV1, ProjectRoot: "/project", Generation: 4,
		LiveID: run.ID, Target: association.TargetV1{PlanID: 2, TaskID: 9}, Revision: 3,
	}
	drift := intelligenceEvent(1, EventError, EventProgress)
	drift.ErrorClass = "scope_mismatch"
	drift.Correlation = EventCorrelation{
		ProjectRoot: "/project", PlanID: 2, TaskID: 9,
		Generation: 4, AssociationRevision: 3,
	}
	if got := DeriveRunIntelligence(run, []Event{drift}); got.State != IntelligencePotentiallyDrifting {
		t.Fatalf("explicit drift intelligence = %#v", got)
	}

	unrelated := intelligenceEvent(2, EventFile, EventCompleted)
	unrelated.Paths = []string{"unrelated/file.go"}
	if got := DeriveRunIntelligence(run, []Event{unrelated}); got.State != IntelligenceWorking {
		t.Fatalf("unrelated file implied drift: %#v", got)
	}

	drift.Correlation.AssociationRevision = 2
	if got := DeriveRunIntelligence(run, []Event{drift}); got.State == IntelligencePotentiallyDrifting {
		t.Fatalf("stale correlation implied current drift: %#v", got)
	}
}

func TestDeriveRunIntelligenceUsesNewestDecisiveEvidence(t *testing.T) {
	blocked := intelligenceEvent(1, EventTool, EventBlocked)
	progress := intelligenceEvent(2, EventCommand, EventProgress)
	got := DeriveRunIntelligence(intelligenceRun(), []Event{progress, blocked})
	if got.State != IntelligenceWorking || len(got.Evidence) != 1 || got.Evidence[0].EventID != progress.ID {
		t.Fatalf("newest evidence intelligence = %#v", got)
	}
}

File: crates/ptrack-agent/src/intelligence_test.rs
use super::{
    Association, AssociationTarget, Event, EventCorrelation, EventKind, EventPhase, Exit,
    IntelligenceConfidence, IntelligenceState, ProcessState, RegistrationKind, Run, RunState,
    Timestamp, derive_run_intelligence,
};

fn run() -> Run {
    Run {
        id: "run-1".to_owned(),
        provider: "codex".to_owned(),
        project_root: "/project".to_owned(),
        terminal_id: "terminal-1".to_owned(),
        registration_kind: RegistrationKind::Launched,
        state: RunState::Running,
        process_state: ProcessState::Running,
        lifecycle_revision: 2,
        ..Run::default()
    }
}

fn event(sequence: u64, kind: EventKind, phase: EventPhase) -> Event {
    Event {
        model_version: 1,
        id: format!("event-{sequence}"),
        run_id: "run-1".to_owned(),
        provider: "codex".to_owned(),
        host_sequence: sequence,
        lifecycle_revision: 2,
        kind,
        phase,
        observed_at: Timestamp::from_unix_nanoseconds(i128::from(sequence) * 1_000_000_000),
        correlation: EventCorrelation {
            project_root: "/project".to_owned(),
            terminal_id: "terminal-1".to_owned(),
            ..EventCorrelation::default()
        },
        ..Event::default()
    }
}

#[test]
fn derivation_uses_newest_current_explicit_evidence() {
    let run = run();
    let waiting = event(2, EventKind::Lifecycle, EventPhase::Waiting);
    let progress = event(3, EventKind::Tool, EventPhase::Progress);
    let stale = Event {
        id: "stale".to_owned(),
        lifecycle_revision: 1,
        phase: EventPhase::Failed,
        ..event(99, EventKind::Lifecycle, EventPhase::Failed)
    };
    let value = derive_run_intelligence(&run, &[progress, stale, waiting]);
    assert_eq!(
        (value.state, value.confidence, value.event_count),
        (
            IntelligenceState::Working,
            IntelligenceConfidence::Medium,
            2
        )
    );
    assert_eq!(value.evidence[0].reason, "recent_observable_activity");
    assert_eq!(value.evidence[0].event_id, "event-3");
}

#[test]
fn failure_completion_and_successful_exit_precedence_is_conservative() {
    let mut run = run();
    run.exit = Some(Exit {
        code: 7,
        result: "failed".to_owned(),
        occurred_at: Timestamp::ZERO,
    });
    assert_eq!(
        derive_run_intelligence(&run, &[]).state,
        IntelligenceState::Failed
    );
    run.exit.as_mut().unwrap().code = 0;
    run.state = RunState::Exited;
    run.process_state = ProcessState::Exited;
    assert_eq!(
        derive_run_intelligence(&run, &[]).state,
        IntelligenceState::Unknown
    );
    assert_eq!(
        derive_run_intelligence(
            &run,
            &[event(1, EventKind::Lifecycle, EventPhase::Completed)]
        )
        .state,
        IntelligenceState::Completed
    );
}

#[test]
fn drift_requires_full_current_association_correlation() {
    let mut run = run();
    run.association = Some(Association {
        version: 1,
        project_root: "/project".to_owned(),
        generation: 5,
        live_id: "run-1".to_owned(),
        target: AssociationTarget {
            plan_id: 2,
            task_id: 9,
        },
        revision: 4,
    });
    let mut mismatch = event(1, EventKind::Error, EventPhase::Progress);
    mismatch.error_class = "scope_mismatch".to_owned();
    assert_eq!(
        derive_run_intelligence(&run, &[mismatch.clone()]).state,
        IntelligenceState::Working
    );
    mismatch.correlation.plan_id = 2;
    mismatch.correlation.task_id = 9;
    mismatch.correlation.generation = 5;
    mismatch.correlation.association_revision = 4;
    assert_eq!(
        derive_run_intelligence(&run, &[mismatch]).state,
        IntelligenceState::PotentiallyDrifting
    );
}

#[test]
fn intelligence_json_exposes_metadata_only_evidence() {
    let value = derive_run_intelligence(
        &run(),
        &[event(1, EventKind::Lifecycle, EventPhase::Waiting)],
    );
    assert_eq!(
        serde_json::to_string(&value).unwrap(),
        r#"{"runId":"run-1","state":"waiting","confidence":"medium","evidence":[{"eventId":"event-1","kind":"lifecycle","phase":"waiting","observedAt":"1970-01-01T00:00:01Z","reason":"explicit_waiting_event"}],"eventCount":1,"lastEventAt":"1970-01-01T00:00:01Z"}"#
    );
}
